pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -237,59 +263,112 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 		glog.Infof("Will report %v as public IP address.", publicAddress)
 	}
 
-	m := &master.Config{
-		Config: &genericapiserver.Config{
-
-			PublicAddress: publicAddress,
-			ReadWritePort: server.SecurePort,
-
-			Authenticator:    originAuthenticator, // this is used to fulfill the tokenreviews endpoint which is used by node authentication
-			Authorizer:       authorizer.NewAlwaysAllowAuthorizer(),
-			AdmissionControl: admissionControl,
-
-			StorageFactory: storageFactory,
-
-			ServiceClusterIPRange: (*net.IPNet)(&server.ServiceClusterIPRange),
-			ServiceNodePortRange:  server.ServiceNodePortRange,
-
-			RequestContextMapper: requestContextMapper,
+	genericConfig := genericapiserver.NewConfig().ApplyOptions(server.GenericServerRunOptions)
+
+	kubeVersion := version.Get()
+	genericConfig.Version = &kubeVersion
+
+	// MUST be in synced with the value in CMServer
+	genericConfig.EnableGarbageCollection = false // disabled until we add the controller
+
+	genericConfig.PublicAddress = publicAddress
+	genericConfig.Authenticator = originAuthenticator // this is used to fulfill the tokenreviews endpoint which is used by node authentication
+	genericConfig.Authorizer = authorizer.NewAlwaysAllowAuthorizer()
+	genericConfig.AdmissionControl = admissionControl
+	genericConfig.RequestContextMapper = requestContextMapper
+	genericConfig.APIResourceConfigSource = getAPIResourceConfig(options)
+	genericConfig.EnableIndex = true
+	genericConfig.OpenAPIConfig = DefaultOpenAPIConfig()
+	genericConfig.SwaggerConfig = genericapiserver.DefaultSwaggerConfig()
+	genericConfig.SwaggerConfig.PostBuildHandler = customizeSwaggerDefinition
+	_, loopbackClientConfig, err := configapi.GetKubeClient(options.MasterClients.OpenShiftLoopbackKubeConfig, options.MasterClients.OpenShiftLoopbackClientConnectionOverrides)
+	if err != nil {
+		return nil, err
+	}
+	genericConfig.LoopbackClientConfig = loopbackClientConfig
+	genericConfig.LegacyAPIGroupPrefixes = LegacyAPIGroupPrefixes
+	genericConfig.SecureServingInfo.BindNetwork = options.ServingInfo.BindNetwork
+	genericConfig.SecureServingInfo.ExtraClientCACerts, err = configapi.GetOAuthClientCertCAs(options)
```

Comment:
whether we pass a cert pool or a list of certificates, I don't want master_config.go pulling specific fields from the config... the helper function GetClientCertCAPool and/or GetClientCertCAs should contain that knowledge